saga/README.md

```diff
@@ -0,0 +1,121 @@
+# Saga Pattern
+
+This example demonstrates how to implement the [Saga pattern](https://microservices.io/patterns/data/saga.html) for realizing distributed transactions across multiple microservices, in a safe and reliable way using change data capture.
+
+Similarly to the outbox pattern, this implementation avoids unsafe dual writes to a service's database and Apache Kafka by channeling all outgoing messages through the originating service's database and capturing them from there using CDC and Debezium.
+
+There are three services involved:
+
+* _order-service:_ originator and orchestrator of the Saga
+* _customer-service:_ approves or rejects the customer's credit needed to fulfill an incoming order
+* _payment-service_ executes the payment associated to an incoming order
+
+## Running the Example
+
+Build and start up:
+
+```console
+$ mvn clean verify
+```
+
+```console
+$ docker-compose up --build
+```
+
+Register the connectors for the different services:
+
+```console
+$ http PUT http://localhost:8083/connectors/order-outbox-connector/config < register-order-connector.json
+$ http PUT http://localhost:8083/connectors/payment-outbox-connector/config < register-payment-connector.json
+$ http PUT http://localhost:8083/connectors/credit-outbox-connector/config < register-credit-connector.json
+```
+
+Place an order:
+
+```console
+$ http POST http://localhost:8080/orders < requests/place-order.json
+```
+
+Examine the emitted messages for `payment` and `credit-approval` in Apache Kafka:
+
+```console
+$ docker run --tty --rm \
+    --network saga-network \
+    debezium/tooling:1.1 \
+    kafkacat -b kafka:9092 -C -o beginning -q \
+    -f "{\"key\":%k, \"headers\":\"%h\"}\n%s\n" \
+    -t payment.request
```

Comment:
if I didn't get the saga flow wrong the credit approval should be executed first so you could swap the order of examining the two topics here in the commands?

Reply:
The order of payment and credit approval event flows doesn't really matter; they could happen in either order, or even interlocked.

Reply:
To close the loop: I've changed the implementation now to ensure a sequential flow: first credit approval, then payment. Updated the readme accordingly. I'm leaving the code for the optimsistic locking in place, as a good measure of caution.